### ISSUE TITLE
The frontend's error dialog should show error `message`'s returned by the backend API

### ISSUE THREAD

**Issue by seanh** [hypothesis/lms#3259]
When the backend API sends a 400 error response with a `message` in the JSON body, the `message` is supposed to be / used to be displayed by the frontend:

https://github.com/hypothesis/lms/blob/1b293991e21659d963fb65c9a05535e6303b8316/lms/views/api/exceptions.py#L39-L44

At some point this got broken. For example if the backend API returns a 400 error response with this body:

```json
{"message": "External request failed"}
```

The frontend currently shows this error dialog:

> ### Something went wrong
> 
> There was a problem fetching this Hypothesis assignment.
> 
> If the problem persists, **send us an email** or **open a support ticket**. You can also visit our **help documents**.
> 
> <kbd>Try again</kbd>

The `"External request failed"` error message is lost, replaced by _There was a problem fetching this Hypothesis assignment_.

**Comment by lyzadanger** [comment]
@seanh Could you be so kind as to provide reproduction instructions for the specific error case you reference in the issue description? I want to be sure I'm testing it throughout this process.

**Comment by lyzadanger** [comment]
@seanh 

I'm having some trouble with:

> 400 with {} in the JSON body: show the authorization dialog (not the error / something-went-wrong dialog)

This seems peculiar and requires the front end to do a bunch of dancing to figure out when an error is an authorization error. Is there a reason an `error_code` couldn't be returned in the response body in this case, even if it's a generic authz error_code? It feels a little fragile at present and  currently requires a convoluted code path that is just begging to break in future. 

Regarding:

> error_code and message should never be returned at the same time. If they are that's a backend bug, but think it'd be fine for the frontend to treat this as case 2: use the error_code and ignore the message.

If the backend wants to impose that exclusive constraint (only one of `error_code` _or_ `message`) that's generally fine, but A) it certainly isn't necessary from the front end's point of view, and B) I don't think the front end should have to enforce it. I'd like the front-end to be able to be dumber, ideally:

* Front end can know to show an authorization-related error dialog if the provided `error_code` indicates it's authz-related (note: in this thread we're discussing errors that are returned in API response bodies. I don't want to confuse things, but there is also authz-related error handling that is done based on error information provided in the config object, when the config indicates the entire app should be in `oauth-redirect-error` "mode". Please try to forget about that for the moment because it doesn't really have bearing on this discussion, but I wanted to point out that there are additional code paths handling authz-related errors of some kind).
* Front end will show _any_ back-end provided `message` and/or `details` if they're present in API response bodies, in all cases, without having to do any reasoning about it

From my perspective, the ideal response structure would be:

```
{
  errorCode: string, // Frontend can switch on this to add any canned text. Ideally always present, even if a default value
  message?: string, // Frontend will always display if present
  details?: object, // Frontend will always display, JSON-stringified, if present
}
```

The only meaningful change here is that `errorCode` is always present.

**Comment by seanh** [comment]
Here's some ways to reproduce the problem.

Unfortunately I don't know of any way to reproduce this problem without hacking the code, since it requires an external API (e.g. Canvas, Blackboard, LTI Outcomes, or h) to fail to respond or to return an error or an invalid or unexpected response.

### Error from the Canvas API

Hack the code like so:

```diff
diff --git a/lms/services/canvas_api/client.py b/lms/services/canvas_api/client.py
index 14556e09..d6c5ecaa 100644
--- a/lms/services/canvas_api/client.py
+++ b/lms/services/canvas_api/client.py
@@ -281,7 +281,7 @@ class CanvasAPIClient:
         # https://canvas.instructure.com/doc/api/files.html#method.files.public_url
 
         return self._client.send(
-            "GET", f"files/{file_id}/public_url", schema=self._PublicURLSchema
+            "GET", f"files/{file_id}/foo", schema=self._PublicURLSchema
         )["public_url"]
 
     class _PublicURLSchema(RequestsResponseSchema):
```

Then launch a Canvas Files assignment. On master the frontend will display _There was a problem fetching this Hypothesis assignment_. It should say _Calling the Canvas API failed_ which is the `message` that the backend returned.

### Error from the Blackboard API

Hack the code like so:

```diff
diff --git a/lms/services/blackboard_api/_schemas.py b/lms/services/blackboard_api/_schemas.py
index 3f609e14..1c9e8300 100644
--- a/lms/services/blackboard_api/_schemas.py
+++ b/lms/services/blackboard_api/_schemas.py
@@ -36,6 +36,7 @@ class BlackboardPublicURLSchema(RequestsResponseSchema):
     """Schema for Blackboard /courses/{courseId}/resources/{resourceId} responses."""
 
     downloadUrl = fields.Str(required=True)
+    foo = fields.Str(required=True)
 
     @post_load
     def post_load(self, data, **_kwargs):  # pylint:disable=no-self-use
```

Then launch a Blackboard Files assignment. On master the frontend will display _There was a problem fetching this Hypothesis assignment._ It should say _External request failed_ which is what the backend sent as the `message`.

### Error from the h API

```diff
diff --git a/lms/services/h_api.py b/lms/services/h_api.py
index 9a9204b8..a94d65e2 100644
--- a/lms/services/h_api.py
+++ b/lms/services/h_api.py
@@ -86,10 +86,14 @@ class HAPI:
         if body is not None:
             request_args["data"] = body
 
+        url = self._base_url + path.lstrip("/")
+        if self._request.url.endswith("/sync"):
+            url = "http://httpbin.org/status/418"
+
         try:
             response = self._http_service.request(
                 method=method,
-                url=self._base_url + path.lstrip("/"),
+                url=url,
                 auth=self._http_auth,
                 headers=headers,
                 **request_args,
```

Instead of _There was a problem fetching this Hypothesis assignment_ it should say _Connecting to Hypothesis failed_.

### Error from the LTI Outcomes API

```diff
diff --git a/lms/services/lti_outcomes.py b/lms/services/lti_outcomes.py
index 51e1bda0..18bfcd3e 100644
--- a/lms/services/lti_outcomes.py
+++ b/lms/services/lti_outcomes.py
@@ -127,7 +127,7 @@ class LTIOutcomesClient:
         """Return the POX body element, checking for errors."""
 
         try:
-            body = data["imsx_POXEnvelopeResponse"]["imsx_POXBody"]
+            body = data["imsx_POXEnvelopeResponse"]["foo"]
             header = data["imsx_POXEnvelopeResponse"]["imsx_POXHeader"]
             status = header["imsx_POXResponseHeaderInfo"]["imsx_statusInfo"][
                 "imsx_codeMajor"
```

Then launch any Moodle assignment as a teacher and select a student from the grading bar. (You may have to launch the assignment as a student first, in order to get a student to appear in the grading bar.) On master I see this, which looks a bit broken:

![Screenshot from 2021-11-08 14-00-37](https://user-images.githubusercontent.com/22498/140756374-511b806c-995c-45ab-9c34-8e9b309f4ea4.png)

**Comment by seanh** [comment]
Opened a separate issue for the visual problems with the error dialog when grading: https://github.com/hypothesis/lms/issues/3352

**Comment by lyzadanger** [comment]
I believe there are three separate things going on at once here, so I'd like to separate them:

1. Whether or not this branch does what it sets out to do, which is to render a backend-provided error message when provided. Based on my own testing previously, and then testing with these provided test cases, _it does_, and I'd like to move forward with this branch and pull it out of draft. _Before these changes_, there were cases that the back-end-provided message would not be rendered at all, _and_ there were cases in which the default error message for all API Errors ("API call failed") might be rendered. This branch fixes those cases.
2. Whether or not the pre-defined user-facing text in [`LaunchErrorDialog`](https://github.com/hypothesis/lms/blob/1d086f65d75a435e3117acd1c254f8c2334dfb32/lms/static/scripts/frontend_apps/components/LaunchErrorDialog.js) should be displayed to users in the `error-fetching` (and possibly others, but I'm not a mind-reader) case. _This text pre-exists these changes_,[^1] so I am going to suggest that any changes that happen in this facet of things be a separate path of work, and may require some input from the original devs (including @marcospri ) who defined these error states to make sure I don't open yet another inadvertent can of worms.
3. Layout issue with error dialogs rendered by the `SubmitGradingForm` component. Pre-exists and is independent of these changes; tracked in #3352 and an easy fix I've already made on another branch, and will open a PR soon.

[^1]: A point of clarity. The changes in this branch will prefer `error.errorMessage` (provided by the back-end) over `error.message` (typically the message set when instantiating a JS Error object). In the case of errors that arise from our back end, we wrap them with a custom JS Error class (`APIError`). In the past, `APIError`'s constructor set a `message` to _either_ the backend-provided message _or_ a non-useful fallback message ("API call failed"). After these changes, `APIError`s all have the same `message`, which is ignored by the component that renders error information to the user. This is what is meant by "prefer `error.errorMessage` to `error.message`" in this PR's title. Users will never see the generic _error object_ message. These changes do not, however, make any logical changes to the canned text that is displayed to users based on `errorCode`, or, in this particular case, a "type" of error encountered during assignment launch.

**Comment by lyzadanger** [comment]
Fix for #3352 https://github.com/hypothesis/lms/pull/3353
